### PR TITLE
Flesh out support for reranking from inline full/lvq vectors while building graph using pq

### PR DESCRIPTION
**Pull request by jbellis**

### CODE REVIEW

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/CachingVectorValues.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ */
+
+package io.github.jbellis.jvector.graph;
+
+import io.github.jbellis.jvector.pq.PQVectors;
+import io.github.jbellis.jvector.vector.types.VectorFloat;
+import org.agrona.collections.Int2ObjectHashMap;
+
+/**
+ * This is NOT a general "make vectors go faster" class.  It is used specificly by diversity computations
```
Comment (jkni) [inline jvector-base/src/main/java/io/github/jbellis/jvector/graph/CachingVectorValues.java]:
```suggestion
 * This is NOT a general "make vectors go faster" class.  It is used specifically by diversity computations
```

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/OnDiskGraphIndexWriter.java [hunk 2]
```diff
@@ -17,20 +17,31 @@
...
+import java.nio.file.Path;
+import java.util.Collection;
 import java.util.EnumMap;
 import java.util.Map;
 import java.util.function.IntFunction;
 
 /**
  * Write a graph index to disk, for later loading as an OnDiskGraphIndex.
+ * <p>
+ * Implements FeatureReader to allow incremental construction of a larger-than-memory graph
```
Comment (jkni) [inline jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/OnDiskGraphIndexWriter.java]: FeatureReader an outdated name? I think you're referring to it working as a feature source

File: jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java [hunk 3]
```diff
@@ -98,116 +102,202 @@ static void runAll(DataSet ds,
...
+                                                                 int efConstruction,
+                                                                 DataSet ds,
+                                                                 Path testDirectory)
+            throws IOException
+    {
+        var floatVectors = ds.getBaseRavv();
+        Map<Set<FeatureId>, GraphIndex> indexes = new HashMap<>();
+        long start;
+        var bsp = BuildScoreProvider.randomAccessScoreProvider(floatVectors, ds.similarityFunction);
+        GraphIndexBuilder builder = new GraphIndexBuilder(bsp, floatVectors.dimension(), M, efConstruction, 1.2f, 1.2f,
```
Comment (jkni) [inline jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java:276]: intentional change from 1.5f, 1.2f here? (compared to building on disk and building in memory prior to change)

File: jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java [hunk 4]
```diff
@@ -98,116 +102,202 @@ static void runAll(DataSet ds,
...
+            var writer = bws.builder.build();
+            writers.put(features, writer);
+            suppliers.put(features, bws.suppliers);
+            if (features.equals(EnumSet.of(FeatureId.LVQ))) {
+                scoringWriter = writer;
+                lvq = (LVQ) bws.builder.getFeature(FeatureId.LVQ);
+            }
+        }
+        if (scoringWriter == null) {
+            throw new IllegalStateException("Incremental compressed builds are only supported by Bench when one of the feature sets is {LVQ}");
```
Comment (jkni) [inline jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java]: out of curiosity, why? Couldn't we do this with full inline vectors?

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/similarity/BuildScoreProvider.java [hunk 5]
```diff
@@ -137,13 +138,14 @@ public SearchScoreProvider.Factory diversityProvider() {
 
     /**
      * Returns a BSP that performs approximate score comparisons using the given PQVectors,
-     * with reranking performed using full resolutions vectors read from the reader
+     * with reranking performed using LVQ PackedVectors
```
Comment (jkni) [inline jvector-base/src/main/java/io/github/jbellis/jvector/graph/similarity/BuildScoreProvider.java]: Couldn't this all be done with full vectors?

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/OnDiskGraphIndexWriter.java (hunk 2 above)
Comment (jbellis) [reply jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/OnDiskGraphIndexWriter.java]: yes, will fix

File: jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java (hunk 3 above)
Comment (jbellis) [reply jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java:276]: i think it's correct as is now, it's supposed to be (1.5, 1.2) for on disk and (1.2, 1.2) for in memory

(retainDiverse is more expensive when we have to fetch from disk, so we allow it to overflow more to call that less frequently)

File: jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java (hunk 4 above)
Comment (jbellis) [reply jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java]: yes, this is purely a Bench limitation not a general one, everything works if you switch LVQ to INLINE_VECTORS

Comment (jbellis) [reply jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java]: I didn't want to try to figure out the right thing to do if user specifies both IV and LVQ

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/similarity/BuildScoreProvider.java (hunk 5 above)
Comment (jbellis) [reply jvector-base/src/main/java/io/github/jbellis/jvector/graph/similarity/BuildScoreProvider.java]: yes, comment is outdated, will update

File: jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java (hunk 4 above)
Comment (jbellis) [reply jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java]: clarified the message

File: jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java (hunk 3 above)
Comment (jkni) [reply jvector-examples/src/main/java/io/github/jbellis/jvector/example/Grid.java:276]: That makes sense.

**Pull request merged**